vignettes/philosophy.Rmd

```diff
@@ -0,0 +1,163 @@
+
+- `underscore_separated` names for functions, parameters and variables.
+- A line of code or comments must be no longer than 80 characters.
+  Pro tip: have RStudio display this margin in the editor.
+  _Tools_ > _Global options_ > _Code_ > _Display_ > _Show margin_
+- Object names must not be longer than 30 characters.
+- Start a new line
+    - after the pipe (`%>%`)
+    - never before but always after `{`
+    - before `}`
+- Use spaces instead of tabs.
+  Pro tip: make RStudio place 2 spaces when you hit the tab key.
+  _Tools_ > _Global options_ > _Code_ > _Editing_ > _Insert spaces for tabs_
+- Use spaces consistently
+    - Use exactly one space before and after
+        - assignments `<-`, `->`, `=`
+        - operators like `+`, `-`,`*`, `/`, ...
+    - No space before and one space after `,`
+    - One space before `(` except in `function(` (don't use `function (`)
+    - One space between `)` and `{`
+    - No space after or before `(` or `[`
+- Use double quotes (`"`) to define character strings.
+- No trailing whitespace
+    - spaces at the end of a line
+    - blank lines at the end of the script
+
+### Static code analysis checks
+
+- Is an object defined before you use it?
+- Do you use an object after you defined it?
+- Use `<-` or `->` to assign something.
+  Only use `=` to pass arguments to a function (e.g. `check_package(fail = TRUE)`).
+- Use `is.na(x)` instead of `x == NA`.
+- Use `seq_len()` or `seq_along()` instead of `1:length(x)`, `1:nrow(x)`, ...
+  When `length(x) == 0`, `1:length(x)` yields `c(1, 0)`, whereas `seq_along(x)` would yield an empty vector.
+- Don't store code in comments.
+  If you don't want to loose code, use version control systems like [`git`](https://git-scm.com/).
+- Avoid code with lots of nested loop or if statements.
+  If the code it too complex, you'll get a warning that the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) is too high.
+  Tips for reducing the code complexity:
+    - Use `assertthat::assert_that()` to validate object or conditions instead of `if() stop()`.
+    - See if you can use `ifelse()` instead of `if()`.
+    - Split the main function of your code over a few sub functions.
+
+## File name conventions
+
+To make this easier to remember we choose the same name conventions for file names as for objects.
+We acknowledge that these rules sometimes clash with requirements from other sources (e.g. `DESCRIPTION` in an R package, `README.md` on GitHub, `.gitignore` for git, ...).
+In such case we allow the file names as required by R, git or GitHub.
+When `check_filename()` does not allow a certain file or folder name and you think this is a false positive, then please open an [issue on GitHub](https://github.com/inbo/checklist/issues) and motivate why this is a false positive.
+
+### Rules for folder names
+
+- Folder names should only contain lower case letters, numbers and underscore (`_`).
+- They can start with a single dot (`.`).
+
```

Comment:
As discussed earlier, I agree with this for R packages, but when using `check_source()` to check source files in a repo, I think we need to be more liberal. Also for file names. Allowing `-` in addition to `_` would do the trick for me (cf https://github.com/inbo/protocols#folder-and-filename-syntaxis).

Reply:
Both functions rely on the same `check_filename()` function. Having the same rules everywhere makes life easier...

Reply:
Please motivate why we should be more liberal.

Reply:
I believe it can be quite common to also have `-` in folder names, possibly in combination with `_`. Currently, only `_` is allowed. 

Examples: 

- https://github.com/inbo/protocols#folder-and-filename-syntaxis
    - folder name is `<protocolcode>_<shorttitle>`: e.g. sfp-401-nl_my-short-title
    - if only `_` is allowed this would become sfp_401_nl_my_short_title, which does not read well and is not regexp friendly (for instance the original `<protocolcode>` and `<shorttitle>` can be easily recovered in the former case by splitting the string at `_`)
- Often folder names contain a date as part of their name. The [ISO 8601 standard](https://nl.wikipedia.org/wiki/ISO_8601) is specified using hyphens: `YYYY-MM-DD`. `YYYYMMDD` is a good alternative, but a lot less readable.
-   If folder or filenames also end up as part of URL's, the general advice is to avoid underscores altogether. See for instance https://x-equals.com/dashes-versus-underscores/ (but I won't go as far as to suggest to only use hyphens)
- Many automatically created file and folder names use hyphens.  For instance: when using `rmarkdown` or `bookdown`, figures are written to a folder named `figures-html` and inside that folder figures are names `chunk-name-1.png`. Another example is `usethis::use_test(foo)`, which will write to a folder `testthat/test-foo.R`
- ...

I believe an extra argument to `check_filename()` that states which word delimiters are allowed (default only `_` underscore; and hyphen or both as the only alternatives) would greatly increase the chance that existing repositories consider using the `checklist` package.  

What do others think about this? @inbo/bmk ?